device/core/src/client_diagnostic.ts

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+
+import { Message, errors } from 'azure-iot-common';
+import { Twin } from './twin';
+import * as dbg from 'debug';
+const debug = dbg('azure-iot-device:Diagnostics');
+
+/**
+ * Data structure to store diagnostic property data
+ */
+export class Diagnostics {
+  // Key of properties in diagnostic context.
+  static DIAGNOSTIC_CONTEXT_PROPERTY: any = {
+    creationTimeUtc: 'creationtimeutc'
+  };
+  // Equal symbol in diagnostic context.
+  static DIAGNOSTIC_CONTEXT_SYMBOL_EQUAL: string = '=';
+  // And symbol in diagnostic context.
+  static DIAGNOSTIC_CONTEXT_SYMBOL_AND: string = '&';
+
+  public id: string;
+  public correlationContext: any;
+
+  /**
+   * @constructor
+   */
+  constructor(id: string, creationTimeUtc: string) {
+    /* Codes_SRS_NODE_DEVICE_DIAGNOSTICS_01_001: [The Diagnostics constructor shall accept a correlation id of message.] */
+    this.id = id;
+    /* Codes_SRS_NODE_DEVICE_DIAGNOSTICS_01_002: [The Diagnostics constructor shall accept a correlation context including creation time of message.] */
+    this.correlationContext = {
+      creationTimeUtc
+    };
+  }
+
+  /**
+   * @method            module:azure-iot-device.Diagnostics.getEncodedCorrelationContext
+   * @description       get the encoded diagnostic correlation context.
+   *
+   * @returns {string}
+   */
+  public getEncodedCorrelationContext(): string {
+    /* Codes_SRS_NODE_DEVICE_DIAGNOSTICS_01_003: [The getEncodedCorrelationContext function returned the encoded string of correlation context.] */
+    let encodedCorrelationContext = '';
+    for (let key in this.correlationContext) {
+      if (encodedCorrelationContext !== '') {
+        encodedCorrelationContext += Diagnostics.DIAGNOSTIC_CONTEXT_SYMBOL_AND;
+      }
+      encodedCorrelationContext += (Diagnostics.DIAGNOSTIC_CONTEXT_PROPERTY[key] + Diagnostics.DIAGNOSTIC_CONTEXT_SYMBOL_EQUAL + this.correlationContext[key]);
+    }
+    return encodedCorrelationContext;
+  }
+}
+
+/**
+ * Handling device client diagnostic information.
+ */
+export class DiagnosticClient {
```

Comment:
>DiagnosticClient [](start = 13, length = 16)

please separate classes in different files - feel free to group the files in a subfolder if you want.